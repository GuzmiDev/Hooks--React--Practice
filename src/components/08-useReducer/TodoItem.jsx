import React from "react";

export const TodoItem = ({ todo, handleToggleTodo, handleDelete, index }) => {
  return (
    <li
      key={todo.id}
      className="list-group-item d-flex justify-content-between align-items-center"
    >
      <p
        onClick={() => {
          handleToggleTodo(todo.id);
        }}
        className={`text-center m-0 pe-auto ${
          todo.done && "text-decoration-line-through"
        }`}
        role="button"
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        onClick={() => {
          handleDelete(todo.id);
        }}
        className="btn btn-danger"
      >
        Borrar
      </button>
    </li>
  );
};
