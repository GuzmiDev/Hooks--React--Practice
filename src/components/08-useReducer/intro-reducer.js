const initialState = [
  {
    id: 5,
    todo: "Hacer ejercicio",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  switch (action.type) {
    case "agregar":
      return [...state, action.payload];
    default:
      return state;
  }
};

let todos = todoReducer();

const newTodo = {
  id: 6,
  todo: "Aprender Angular",
  done: true,
};

const action = {
  type: "agregar",
  payload: newTodo,
};

todos = todoReducer(initialState, action);

console.log(todos);
