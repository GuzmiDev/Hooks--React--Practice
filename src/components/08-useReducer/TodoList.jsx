import React from "react";
import { TodoItem } from "./TodoItem";

export const TodoList = ({ todos, handleDelete, handleToggleTodo }) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((todo, i) => (
        <TodoItem
          todo={todo}
          handleDelete={handleDelete}
          handleToggleTodo={handleToggleTodo}
          key={todo.id}
          index={i}
        />
      ))}
    </ul>
  );
};
