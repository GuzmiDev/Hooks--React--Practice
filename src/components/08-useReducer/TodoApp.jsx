import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import "./styles.css";
import { TodoAdd } from "./TodoAdd";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  const handleDelete = (todoId) => {
    dispatch({
      type: "delete",
      payload: todoId,
    });
  };

  const handleToggleTodo = (todoId) => {
    dispatch({
      type: "done",
      payload: todoId,
    });
  };

  return (
    <div>
      <h1 className="text-center">TodoApp({todos.length}) </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggleTodo={handleToggleTodo}
          />
        </div>

        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
};
