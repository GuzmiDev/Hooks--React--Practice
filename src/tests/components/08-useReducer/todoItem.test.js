import { shallow } from "enzyme";
import { TodoItem } from "../../../components/08-useReducer/TodoItem";
import { demosTodos } from "../../fixtures/demoTodos";

describe("Pruebas en <TodoItem />", () => {
  const handleToggleTodo = jest.fn();
  const handleDelete = jest.fn();
  let wrapper = shallow(
    <TodoItem
      todo={demosTodos[0]}
      index={0}
      handleDelete={handleDelete}
      handleToggleTodo={handleToggleTodo}
    />
  );

  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de llamar la función handleDelete", () => {
    wrapper.find("button").simulate("click");

    expect(handleDelete).toHaveBeenCalledWith(demosTodos[0].id);
  });

  test("debe de llamar la función handleToggle", () => {
    wrapper.find("p").simulate("click");

    expect(handleToggleTodo).toHaveBeenCalledWith(demosTodos[0].id);
  });

  test("debe de mostrar el texto correctamente", () => {
    expect(wrapper.find("p").text().trim()).toBe(`1. ${demosTodos[0].desc}`);
  });

  test("debe de mostrar la clase text-decoration-line-thoruh", () => {
    wrapper = shallow(<TodoItem todo={{ ...demosTodos[0], done: true }} />);

    expect(
      wrapper.find("p").hasClass("text-decoration-line-through")
    ).toBeTruthy();
  });
});
