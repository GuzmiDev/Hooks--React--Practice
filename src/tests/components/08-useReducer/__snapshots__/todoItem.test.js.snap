// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoItem /> debe de mostrarse correctamente 1`] = `
<li
  className="list-group-item d-flex justify-content-between align-items-center"
  key="1"
>
  <p
    className="text-center m-0 pe-auto false"
    onClick={[Function]}
    role="button"
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
